Sources/SwiftParser/Declarations.swift

```diff
@@ -1371,7 +1371,7 @@ extension Parser {
     var effectSpecifiers = self.parseDeclEffectSpecifiers()
 
     let output: RawReturnClauseSyntax?
-    if self.at(.arrow) {
+    if let _ = self.canRecoverTo(TokenSpec(.arrow)) {
```

Comment:
The general style in this code base is the following if we aren't using the result of `canRecoverTo`
```suggestion
    if self.canRecoverTo(TokenSpec(.arrow)) != nil {
```

same for `{`

---

Also, can you make sure that the arrow is on the same line the current token by passing `allowAtStartOfLine: false` to the `TokenSpec`? I think it's not as much of an issue here but I've commented on the `{` why this check is important for correct parsing.